File: hs/misc/Priority.hs
module Priority where

import Data.List (sort, sortBy, elemIndex)
import Data.Maybe (fromJust)
import Data.Time (toGregorian, fromGregorian, localDay,
                  getZonedTime, zonedTimeToLocalTime)
import Data.Time.Calendar (diffDays)
import Control.Monad (forM, mapM_)

weights :: [(String, Float)]
weights = [ ("Bid",       1.0),
            ("LeaveRequest", 0.1)
          ]

data Task = Task {
    priority   :: Float,
    kind       :: String,
    weight     :: Float,
    value      :: Float,
    due        :: (Integer, Int, Int),
    importance :: Float,
    urgency    :: Float

} deriving (Eq, Show)

records :: [(String, Float, (Integer, Int, Int))]
records =
    [ ("Bid",       100.0,  (2009, 3, 1))
    , ("Bid",       1000.0, (2009, 4, 1))
    , ("Bid",       100.0,  (2009, 5, 1))
    , ("Bid",       100.0,  (2009, 6, 1))
    , ("LeaveRequest", 100.0,  (2009, 7, 1))
    , ("Bid",       100.0,  (2009, 8, 1))
    , ("Bid",       100.0,  (2009, 9, 1))
    , ("Bid",       100.0,  (2009, 10,1))
    ]

process :: [(String, Float, (Integer, Int, Int))] -> IO [Task]
process records = do
    let total = sum [value | (_, value, _) <- records]
    due_days' <- mapM daysTo [date | (_, _, date) <- records]
    let due_days = reverse $ sort due_days'
    tasks <- forM records $ \record -> do
        let (kind, value, due_date) = record
        days <- daysTo due_date
        let weight = fromJust $ lookup kind weights
            relative_importance = value / (total + value)
            relative_urgency = percentrank due_days days
            task = Task {
                priority = weight * relative_importance + relative_urgency,
                kind = kind,
                weight = weight,
                value = value,
                due = due_date,
                importance = relative_importance,
                urgency = relative_urgency
            }
        return [task]
    return (concat tasks)

daysTo :: (Integer, Int, Int) -> IO Integer
daysTo (year, month, day) = do
    now <- getZonedTime
    let (y, m, d) = toGregorian (localDay (zonedTimeToLocalTime now))
        current = fromGregorian y m d
        prior = fromGregorian year month day
    return (diffDays prior current)

percentrank :: (Fractional b, Eq a) => [a] -> a -> b
percentrank xs x = rank / (n-1)
    where
        n = fromIntegral (length xs)
        rank = fromIntegral (fromJust $ elemIndex x xs)

prioritize :: [Task] -> [Task]
prioritize tasks = reverse $ sortBy compare tasks
    where
        compare Task {priority = p1}
                Task {priority = p2}
            | p1 < p2 = LT
            | otherwise = GT

main = do
    tasks <- process records
    let sorted = prioritize tasks
    mapM_ (putStrLn . show) sorted

File: py/misc/priority.py
from datetime import date

weights = {
    'Bid'      : 1.0,
    'LeaveRequest': 0.1
}

rnd = lambda x: round(x,2)

class Task(object):
    """a simple Task class"""

    def __init__(self, kind, value, due_date=None):
        self.kind = kind
        self.value = value
        self.due_date = due_date if due_date else date.today()
        self.weight = weights[kind]
        self.urgency = 0.0
        self.importance = 0.0

    def __repr__(self):
        return '<Task: %5s - %4s - %5s - %12s - %3s - %s>' % (
            rnd(self.priority), rnd(self.importance), rnd(self.urgency),
            self.kind, self.days_to, self.value)

    def __cmp__(self, other):
        return cmp(self.priority, other.priority)

    @property
    def days_to(self):
        return (self.due_date - date.today()).days

    @property
    def priority(self):
        return self.weight * (self.importance + self.urgency)

    def relative_urgency(self, N, due_days):
        rank = due_days.index(self.days_to)
        return float(rank) / (N - 1)

    def relative_importance(self, total_value):
        return self.value / total_value

    @staticmethod
    def prioritize(tasks):
        print ("<Task: priority - importance - urgency - "
               "kind - days_to - value>\n")
        N = len(tasks)
        total_value = sum(t.value for t in tasks)
        due_days = sorted([t.days_to for t in tasks], reverse=True)
        for i in tasks:
            i.importance = i.relative_importance(total_value)
            i.urgency = i.relative_urgency(N, due_days)

        for i in sorted(tasks, reverse=True):
            print i

tasks = [
    # name               value       due_date
    Task("Bid",       1000000.0,  date(2009,4,1)),
    Task("Bid",       500400.0,   date(2009,5,1)),
    Task("Bid",       1000000.0,  date(2009,6,1)),
    Task("LeaveRequest", 0.0,        date(2009,7,1)),
    Task("LeaveRequest", 0.0,        date(2009,8,1)),
]


if __name__ == '__main__':
    Task.prioritize(tasks)
